contracts/utils/math/Math.sol

```diff
@@ -537,41 +537,82 @@ library Math {
-            value >>= exp;
-            result += exp;
-
-            exp = 64 * SafeCast.toUint(value > (1 << 64) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 32 * SafeCast.toUint(value > (1 << 32) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 16 * SafeCast.toUint(value > (1 << 16) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 8 * SafeCast.toUint(value > (1 << 8) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 4 * SafeCast.toUint(value > (1 << 4) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 2 * SafeCast.toUint(value > (1 << 2) - 1);
-            value >>= exp;
-            result += exp;
+            // Round `x` down to the closest power of 2 using the Seander's algorithm.
+            // Reference: https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+            x |= x >> 1;
+            x |= x >> 2;
+            x |= x >> 4;
+            x |= x >> 8;
+            x |= x >> 16;
+            x |= x >> 32;
+            x |= x >> 64;
+            x |= x >> 128;
+            x = (x >> 1) + 1;
+
+            uint256 prod0;
+            assembly {
+                // 1. Compute `n ≡ x mod 255`, because `x` is power of two, the resulting `n` can only be one of the
+                // following values: 1, 2, 4, 8, 16, 32, 64 or 128.
+                let n := mod(x, 255)
+
+                // 2. Compute `prod0 = log2(n)` using the OPCODE BYTE to lookup a 256-bit word, which we use as table. Notice the
+                // maximum size of our table is 32, but their values can be greater than 32, so first we map `n` into an unique
+                // index between 0~31 by computing `n % 11`, this maps all `n` values to an unique index, as demonstrated below.
+
+                // | n ≡ x mod 255 | index ≡ n mod 11 |    prod0 = log2(n)   |
+                // +-+-+-+-+-+-+-+-+-+-+-+-++-+-+-+-+-+-+-+-+-+-+-+-++-+-+-+-+
+                // |        1      |    1 % 11 == 1   |     log2(1) == 0     |
+                // |        2      |    2 % 11 == 2   |     log2(2) == 1     |
+                // |        4      |    4 % 11 == 4   |     log2(4) == 2     |
+                // |        8      |    4 % 11 == 8   |     log2(8) == 3     |
+                // |       16      |   16 % 11 == 5   |    log2(16) == 4     |
+                // |       32      |   32 % 11 == 10  |    log2(32) == 5     |
+                // |       64      |   64 % 11 == 9   |    log2(64) == 6     |
+                // |      128      |  128 % 11 == 7   |   log2(128) == 7     |
+                let index := mod(n, 11)
+
+                // Below we perform the table lookup, the table stores the result of `log2(n)`
+                // in the corresponding byte at `index`.
+                prod0 := byte(index, 0x0000010002040007030605000000000000000000000000000000000000000000)
```

Comment:
would that work? In would use a `push11` instead of `push32`, and reduce the bytecode size/deployment cost
```suggestion
                prod0 := byte(index, 0x0000010002040007030605)
```

Reply:
Confirmed, using Solc `0.8.27` when set optimizer runs below `600`, it outputs:
```
PUSH9 0x010002040007030605 PUSH1 0xa8 SHL
```
But for any value equal or greater than `600` optimizer runs, solidity seems to prioritize execution cost:
```
PUSH30 0x010002040007030605000000000000000000000000000000000000000000
```

Code:
```solidity
pragma solidity >=0.7.0 <0.9.0;

contract Log2 {
    fallback() payable external {
        uint256 x;
        assembly {
            x := calldataload(0)
        }
        unchecked {
            x |= x >> 1;
            x |= x >> 2;
            x |= x >> 4;
            x |= x >> 8;
            x |= x >> 16;
            x |= x >> 32;
            x |= x >> 64;
            x |= x >> 128;
            x = (x >> 1) + 1;
            uint256 prod0;
            {
                uint256 n = x % 255;
                uint256 index = n % 11;
                assembly {
                    prod0 := byte(index, 0x0000010002040007030605000000000000000000000000000000000000000000)
                }
            }
            uint256 prod1 = x >> prod0;
            prod1 *= 0x0008101820283038404850586068707880889098a0a8b0b8c0c8d0d8e0e8f0f8;
            prod1 >>= 248;
            assembly {
                mstore(0, add(prod0, prod1))
                return (0, 0x20)
            }
        }
    } 
}
```
Obs: when comparing algorithms I like to use `fallback() payable external` to remove the selector overhead from the final bytecode, makes it easier to analyze. For example if I don't manually scope `n` and `index`, solidity outputs worst code than without manual scoping.